accord-core/src/main/java/accord/utils/SortedArrays.java

```diff
@@ -0,0 +1,535 @@
+    {
+        A[] result;
+        int resultCount;
+        int ai = 0, ri = 0;
+        while (true)
+        {
+            long ari = findNextIntersection(slice, ai, select, ri, cmp1, cmp2, Search.CEIL);
+            if (ari < 0)
+            {
+                if (ai == slice.length)
+                    return slice;
+
+                return Arrays.copyOf(slice, ai);
+            }
+
+            int nextai = (int)(ari >>> 32);
+            if (ai != nextai)
+            {
+                resultCount = ai;
+                result = factory.apply(ai + (slice.length - nextai));
+                System.arraycopy(slice, 0, result, 0, resultCount);
+                ai = nextai;
+                break;
+            }
+
+            ri = (int)ari;
+            ai = exponentialSearch(slice, nextai, slice.length, select[ri], cmp2, Search.FLOOR) + 1;
+        }
+
+        while (true)
+        {
+            int nextai = exponentialSearch(slice, ai, slice.length, select[ri], cmp2, Search.FLOOR) + 1;
+            while (ai < nextai)
+                result[resultCount++] = slice[ai++];
+
+            long ari = findNextIntersection(slice, ai, select, ri, cmp1, cmp2, Search.CEIL);
+            if (ari < 0)
+            {
+                if (resultCount < result.length)
+                    result = Arrays.copyOf(result, resultCount);
+
+                return result;
+            }
+
+            ai = (int)(ari >>> 32);
+            ri = (int)ari;
+        }
+    }
+
+    /**
+     * Copy-on-write insert into the provided array; returns the same array if item already present, or a new array
+     * with the item in the correct position if not. Linear time complexity.
+     */
+    public static <T extends Comparable<? super T>> T[] insert(T[] src, T item, IntFunction<T[]> factory)
+    {
+        int insertPos = Arrays.binarySearch(src, item);
+        if (insertPos >= 0)
+            return src;
+        insertPos = -1 - insertPos;
+
+        T[] trg = factory.apply(src.length + 1);
+        System.arraycopy(src, 0, trg, 0, insertPos);
+        trg[insertPos] = item;
+        System.arraycopy(src, insertPos, trg, insertPos + 1, src.length - insertPos);
+        return trg;
+    }
+
+    /**
+     * Equivalent to {@link Arrays#binarySearch}, only more efficient algorithmically for linear merges.
+     * Binary search has worst case complexity {@code O(n.lg n)} for a linear merge, whereas exponential search
+     * has a worst case of {@code O(n)}. However compared to a simple linear merge, the best case for exponential
+     * search is {@code O(lg(n))} instead of {@code O(n)}.
+     */
+    public static <T1, T2 extends Comparable<? super T1>> int exponentialSearch(T1[] in, int from, int to, T2 find)
+    {
+        return exponentialSearch(in, from, to, find, Comparable::compareTo, Search.FAST);
+    }
+
+    // TODO: check inlining elides this
+    public enum Search { FLOOR, CEIL, FAST }
```

Comment:
Spoke in slack, can we add docs for this?  This is similar to `java.util.NavigableMap#floorEntry` and `java.util.NavigableMap#ceilingEntry` but would be good if the reader can read the docs rather than the code usage.